src/Visitor/BuildColumnsVisitor.php

```diff
@@ -204,4 +210,14 @@ protected function parseValue($value)
 
         return $value;
     }
+
+    protected function parseMappedRuleValue(QuerySymbol $query, ColumnRule $rule)
+    {
+        if (! $rule->map->offsetExists($query->value)) {
+            throw new InvalidSearchStringException();
+        }
+
+        $query = new QuerySymbol($query->key, $query->operator, $rule->map[$query->value]);
+        return $this->buildBasicQuery($query, $rule);
+    }
```

Comment:
I don't think we should force values to be part of the map. If we do then we limit ourselves to the usage of a closed enum.

However, if we don't, we add the ability for people to create aliases for certain values.

Therefore I suggest having the `parseMappedRuleValue` only do the parsing and returning a new query symbol with the mapped value.

Something like that:

```php
protected function parseMappedRuleValue(QuerySymbol $query, ColumnRule $rule)
{
    if (! $rule->map || ! $rule->map->has($query->value)) {
        return $query;
    }

    return new QuerySymbol($query->key, $query->operator, $rule->map->get($query->value);
}
```

Which is simply called like this:

```diff
- if ($rule && $rule->map && $rule->map->isNotEmpty()) {
-     return $this->parseMappedRuleValue($query, $rule);
- }
+ $query = $this->parseMappedRuleValue($query, $rule);
```

And therefore there is no need to throw an `InvalidSearchStringException` during this process.

What do you think?

Reply:
@lorisleiva i stand by my implementation. I believe we should only allow for closed enums. 

The reason being a security concern. I does not really makes sense to me that given this map:

```
['support_level_id' => ['key' => 'support_level', 'map' => [
    'testing' => 1,
    'community' => 2,
    'official' => 3
]]]
```

And searching like: `support_level:4` doesn't throw a parse value error for enum types because the given value does not exists. Enums for me is an explicit defined list of values that i would allow the user to search using. If we don't use closed enums then we expose the user the ability to parse in any value that's not apart of the enum map which i believe defeats purpose of them. 

But instead of arguing with you about this i propose we add another possible option to the rule maybe called `strict_map` which is a boolean which could alternate the behavour of this. This is after all a package so we have many different opinions.